File: tests/def+fn.out.py
class Loki:
    @staticmethod
    def printf(x):
        print(x)
    @staticmethod
    def plus(*args):
        return reduce((lambda x, y : x + y), args)
    @staticmethod
    def minus(*args):
        return reduce((lambda x, y : x - y), args)
    @staticmethod
    def div(*args):
        return reduce((lambda x, y : x / y), args)
    @staticmethod
    def mult(*args):
        return reduce((lambda x, y : x * y), args)
    @staticmethod
    def and_(*args):
        return reduce((lambda x, y : x and y), args)
    @staticmethod
    def or_(*args):
        return reduce((lambda x, y : x or y), args)
    @staticmethod
    def eq(*args):
        return not not reduce((lambda x, y : x if x == y else False), args)
    @staticmethod
    def neq(*args):
        return not Loki.eq(*args)
    @staticmethod
    def lt(*args):
        return not not reduce((lambda x, y : y if x < y else False), args)
    @staticmethod
    def lte(*args):
        return not not reduce((lambda x, y : y if x <= y else False), args)
    @staticmethod
    def gt(*args):
        return not not reduce((lambda x, y : y if x > y else False), args)
    @staticmethod
    def gte(*args):
        return not Loki.lt(*args)
    @staticmethod
    def mod(x, y):
        return x % y
    @staticmethod
    def range(n):
        return range(n)
    @staticmethod
    def get(e, i):
        return e[i]
    @staticmethod
    def set(x, v):
        x = v
    @staticmethod
    def assoc(x, i, v):
        x[i] = v
        return x
    @staticmethod
    def in_(x, l):
        return (x in l)
    @staticmethod
    def sc(n, x, l):
        return n[x:l]
    @staticmethod
    def dc(n, x, l):
        return n[x::l]
    @staticmethod
    def dcm(n, x, m, l):
        return n[x:m:l]
    @staticmethod
    def not_ (x):
        return not x

#END LOKI HELPER FUNCTIONS
x = None
y = 2
x = y
x = Loki.plus(1, 2)
y = [1, 2]
x = None
x = None
x = None
x = 1
x = 1
x = 1
def y (x):
    return Loki.printf(x)
def z ():
    return Loki.printf("hello")
def a ():
    return []
def y (x):
    return Loki.printf(x)
def y (x):
    return Loki.printf(x)
def y (x):
    return Loki.printf(x)
Loki.printf((lambda  : [[]])())
Loki.printf((lambda x : [[]])(1))
Loki.printf((lambda x : [Loki.printf(x)])(4))
Loki.printf((lambda x, y : [[]])(2, 5))
Loki.printf((lambda x, y : [Loki.printf(Loki.plus(x, y))])(1, 9))
Loki.printf((lambda x, y, z : [Loki.printf("first"), Loki.printf(Loki.plus(x, y, z))])(1, 1, 1))
Loki.printf((lambda x : [Loki.printf(x)])(7))
Loki.printf((lambda x : [Loki.printf(x)])(12))

File: tests/def+fn.out.js
var loki = (function (){
    var loki = {};

    var sliceArgs = function (args, start) {
        start = typeof start !== 'undefined' ? start : 0;
        return args.length > start ? [].slice.call(args, start) : [];
    };

    var assertIsFunction = function (f) {
        if (typeof f !== "function")
        {throw "loki error: curry expected a function"}
        return f;
    };

    var curry = function (fn) {
        assertIsFunction(fn);
        return function inner() {
            var _args = sliceArgs(arguments);
            if (_args.length === fn.length) {
                return fn.apply(null, _args);
            } else if (_args.length > fn.length) {
                var initial = fn.apply(null, _args);
                return foldl(fn, initial, _args.slice(fn.length));
            } else {
                return function() {
                    var args = sliceArgs(arguments);
                    return inner.apply(null, _args.concat(args));
                };
            }
        };
    };

    var each = curry(function (iterator, items) {
        assertIsFunction(iterator);
        if (items == null || !Array.isArray(items)) {return;}
        items.forEach(function (e, i) {iterator.call(null, e, i)});
    });

    var foldl = curry(function (iterator, acc, xs) {
        assertIsFunction(iterator);
        each(function (x, i) {
            acc = iterator.call(null, acc, x, i);
        }, xs);
        return acc;
    });

    loki.extend = function (destination, source) {
        for (var k in source) {
            if (source.hasOwnProperty(k)) {
                destination[k] = source[k];
            }
        }
        return destination;
    }

    loki.print = function() {
        var _log = function(x) {
            if (typeof console === "object") {console.log(x);}
            else {assertIsFunction(print)(x);}
        };
        var args = sliceArgs(arguments);
        args.forEach(_log);
    };

    loki.get = function(e, i) {return e[i];};
    loki.set = function(x, v) {x = v;};
    loki.assoc = function(x, i, v) {x[i] = v;return x};
    loki.range = function(N) {return Array.apply(null, {length: N}).map(Number.call, Number);};

    //Arithmetic
    loki.plus  = curry(function(x, y) {return x + y});
    loki.minus = curry(function(x, y) {return x - y});
    loki.mult  = curry(function(x, y) {return x * y});
    loki.div   = curry(function(x, y) {return x / y});
    loki.mod   = curry(function(x, y) {return x % y});

    //Logic
    loki.and = curry(function(x, y) {return x && y});
    loki.or  = curry(function(x, y) {return x || y});
    loki.eq  = curry(function(x, y) {return x === y});
    loki.neq = curry(function(x, y) {return x !== y});
    loki.lt  = curry(function(x, y) {return x < y});
    loki.lte = curry(function(x, y) {return x <= y});
    loki.gt  = curry(function(x, y) {return x > y});
    loki.gte = curry(function(x, y) {return x >= y});

    return loki;
})();
//END LOKI HELPER FUNCTIONS
var x;
var y = 2;
var x = y;
var x = loki.plus(1, 2);
var y = [1, 2];
var x;
var x;
var x;
var x = 1;
var x = 1;
var x = 1;
var y = function (x) {
return loki.print(x)
};
var z = function () {
return loki.print("hello")
};
var a = function () {
return []
};
var y = function (x) {
return loki.print(x)
};
var y = function (x) {
return loki.print(x)
};
var y = function (x) {
return loki.print(x)
};
loki.print(function () {
return []
}());
loki.print(function (x) {
return []
}(1));
loki.print(function (x) {
return loki.print(x)
}(4));
loki.print(function (x, y) {
return []
}(2, 5));
loki.print(function (x, y) {
return loki.print(loki.plus(x, y))
}(1, 9));
loki.print(function (x, y, z) {
loki.print("first");
return loki.print(loki.plus(x, y, z))
}(1, 1, 1));
loki.print(function (x) {
return loki.print(x)
}(7));
loki.print(function (x) {
return loki.print(x)
}(12));